File: examples/fastapi-app/app/handlers/ecommerce.py
"""E-commerce order processing step handlers.

5 sequential steps demonstrating a linear pipeline:
  ValidateCart -> ProcessPayment -> UpdateInventory -> CreateOrder -> SendConfirmation

Thin DSL wrappers that delegate to app.services.ecommerce for business logic.
"""

from __future__ import annotations

from tasker_core.step_handler.functional import depends_on, inputs, step_handler
from tasker_core.types import StepContext

from app.services import ecommerce as svc
from app.services.types import (
    EcommerceCreateOrderResult,
    EcommerceOrderInput,
    EcommerceProcessPaymentResult,
    EcommerceUpdateInventoryResult,
    EcommerceValidateCartResult,
)


@step_handler("validate_cart")
@inputs(EcommerceOrderInput)
def validate_cart(inputs: EcommerceOrderInput, context: StepContext):
    return svc.validate_cart_items(inputs.resolved_items)


@step_handler("process_payment")
@depends_on(cart_result=("validate_cart", EcommerceValidateCartResult))
@inputs(EcommerceOrderInput)
def process_payment(
    cart_result: EcommerceValidateCartResult,
    inputs: EcommerceOrderInput,
    context: StepContext,
):
    return svc.process_payment(
        payment_token=inputs.payment_token,
        total=cart_result.total or 0.0,
    )


@step_handler("update_inventory")
@depends_on(cart_result=("validate_cart", EcommerceValidateCartResult))
def update_inventory(cart_result: EcommerceValidateCartResult, context: StepContext):
    return svc.update_inventory(cart_result.validated_items or [])


@step_handler("create_order")
@depends_on(
    cart_result=("validate_cart", EcommerceValidateCartResult),
    payment_result=("process_payment", EcommerceProcessPaymentResult),
    inventory_result=("update_inventory", EcommerceUpdateInventoryResult),
)
@inputs(EcommerceOrderInput)
def create_order(
    cart_result: EcommerceValidateCartResult,
    payment_result: EcommerceProcessPaymentResult,
    inventory_result: EcommerceUpdateInventoryResult,
    inputs: EcommerceOrderInput,
    context: StepContext,
):
    return svc.create_order(
        cart=cart_result,
        payment=payment_result,
        inventory=inventory_result,
        customer_email=inputs.customer_email,
    )


@step_handler("send_confirmation")
@depends_on(order_result=("create_order", EcommerceCreateOrderResult))
@inputs(EcommerceOrderInput)
def send_confirmation(
    order_result: EcommerceCreateOrderResult,
    inputs: EcommerceOrderInput,
    context: StepContext,
):
    return svc.send_confirmation(order=order_result, customer_email=inputs.customer_email)

File: examples/bun-app/src/handlers/ecommerce.ts
/**
 * E-commerce order processing step handlers.
 *
 * 5 sequential steps demonstrating a linear pipeline:
 *   ValidateCart -> ProcessPayment -> UpdateInventory -> CreateOrder -> SendConfirmation
 *
 * Thin DSL wrappers that delegate to ../services/ecommerce for business logic.
 */

import { defineHandler } from "@tasker-systems/tasker";
import { z } from "zod";
import * as svc from "../services/ecommerce";
import type {
	EcommerceCreateOrderResult,
	EcommerceProcessPaymentResult,
	EcommerceUpdateInventoryResult,
	EcommerceValidateCartResult,
	PaymentInfo,
} from "../services/schemas";
import { CartItemSchema } from "../services/schemas";

export const ValidateCartHandler = defineHandler(
	"Ecommerce.StepHandlers.ValidateCartHandler",
	{ inputs: { cartItems: "cart_items" } },
	async ({ cartItems }) =>
		svc.validateCartItems(z.array(CartItemSchema).optional().parse(cartItems)),
);

export const ProcessPaymentHandler = defineHandler(
	"Ecommerce.StepHandlers.ProcessPaymentHandler",
	{
		depends: { cartResult: "validate_cart" },
		inputs: { paymentInfo: "payment_info" },
	},
	async ({ cartResult, paymentInfo }) =>
		svc.processPayment(
			cartResult as EcommerceValidateCartResult,
			paymentInfo as PaymentInfo | undefined,
		),
);

export const UpdateInventoryHandler = defineHandler(
	"Ecommerce.StepHandlers.UpdateInventoryHandler",
	{ depends: { cartResult: "validate_cart" } },
	async ({ cartResult }) =>
		svc.updateInventory(
			(cartResult as EcommerceValidateCartResult).validated_items,
		),
);

export const CreateOrderHandler = defineHandler(
	"Ecommerce.StepHandlers.CreateOrderHandler",
	{
		depends: {
			cartResult: "validate_cart",
			paymentResult: "process_payment",
			inventoryResult: "update_inventory",
		},
		inputs: { customerEmail: "customer_email" },
	},
	async ({ cartResult, paymentResult, inventoryResult, customerEmail }) =>
		svc.createOrder(
			cartResult as EcommerceValidateCartResult,
			paymentResult as EcommerceProcessPaymentResult,
			inventoryResult as EcommerceUpdateInventoryResult,
			customerEmail as string | undefined,
		),
);

export const SendConfirmationHandler = defineHandler(
	"Ecommerce.StepHandlers.SendConfirmationHandler",
	{
		depends: {
			orderResult: "create_order",
			cartResult: "validate_cart",
		},
		inputs: { customerEmail: "customer_email" },
	},
	async ({ orderResult, cartResult, customerEmail }) =>
		svc.sendConfirmation(
			orderResult as EcommerceCreateOrderResult,
			cartResult as EcommerceValidateCartResult | undefined,
			customerEmail as string | undefined,
		),
);
